src/mttt4.c

```diff
@@ -0,0 +1,1013 @@
+    int i;
+    POSITION position = 0;
+
+
+    for(i = 0; i < BOARDSIZE; i++)
+        position += g3Array[i] * (int)theBlankOX[i]; /* was (int)position... */
+
+
+    return(position);
+}
+
+
+/************************************************************************
+**
+** NAME:        FourOfTheSame
+**
+** DESCRIPTION: Return TRUE iff there are three-in-a-row.
+**
+** INPUTS:      BlankOX theBlankOX[BOARDSIZE] : The BlankOX array.
+**              int a,b,c                     : The 3 positions to check.
+**
+** OUTPUTS:     (BOOLEAN) TRUE iff there are three-in-a-row.
+**
+************************************************************************/
+
+
+BOOLEAN FourOfTheSame(BlankOX *theBlankOX, int a, int b, int c, int d) {
+    return(       theBlankOX[a] == theBlankOX[b] &&
+                  theBlankOX[b] == theBlankOX[c] &&
+                  theBlankOX[c] == theBlankOX[d] &&
+                  theBlankOX[d] != Blank );
+}
+
+
+/************************************************************************
+**
+** NAME:        AllFilledIn
+**
+** DESCRIPTION: Return TRUE iff all the blanks are filled in.
+**
+** INPUTS:      BlankOX theBlankOX[BOARDSIZE] : The BlankOX array.
+**
+** OUTPUTS:     (BOOLEAN) TRUE iff all the blanks are filled in.
+**
+************************************************************************/
+
+
+BOOLEAN AllFilledIn(BlankOX *theBlankOX) {
+    BOOLEAN answer = TRUE;
+    int i;
+
+
+    for(i = 0; i < BOARDSIZE; i++)
+        answer &= (theBlankOX[i] == o || theBlankOX[i] == x);
+
+
+    return(answer);
+}
+
+
+/************************************************************************
+**
+** NAME:        WhoseTurn
+**
+** DESCRIPTION: Return whose turn it is - either x or o. Since x always
+**              goes first, we know that if the board has an equal number
+**              of x's and o's, that it's x's turn. Otherwise it's o's.
+**
+** INPUTS:      BlankOX theBlankOX : The input board
+**
+** OUTPUTS:     (BlankOX) : Either x or o, depending on whose turn it is
+**
+************************************************************************/
+
+
+BlankOX WhoseTurn(BlankOX *theBlankOX) {
+    int i, xcount = 0, ocount = 0;
+
+
+    for(i = 0; i < BOARDSIZE; i++)
+        if(theBlankOX[i] == x)
+            xcount++;
+        else if(theBlankOX[i] == o)
+            ocount++;
+        /* else don't count blanks */
+
+
+    if(xcount == ocount)
+        return(x);  /* in our TicTacToe, x always goes first */
+    else
+        return(o);
+}
+
+
+int NumberOfOptions() {
+    return 2;
+}
+ 
+
+
+int getOption()
+{
+    int option = 0;
+    option += gStandardGame;
+    option *= 2;
+    option += (int)Diamond;
+    return option+1;
+}
+
+
+void setOption(int option)
+{
+    option -= 1;
+    Diamond = (BOOLEAN) (option % 2);
+    option /= 2;
+    gStandardGame = option;
+}
+
+
+POSITION ActualNumberOfPositions(int variant) {
+    (void)variant;
+    return 3287691;
+}

```

Comment:
umm but I don't know how I would calculate that

Reply:
so that means you shouldnt?
ive never been fast or accurate with combinatorics, so maybe im thinking incorrectly... it could be same number if all positions are still reachable. proofs on that would be sufficient, but i dont know if we can achieve that in time

Reply:
So, this is a very good upperbound, because this number was not calculated using combinatorics, but it's like the actual possible positions it can have during the 4 by 4 game execution.. I have the ability to find it for diamond variant, but I'm afraid I'm running out of time. I will comment it, but I'll Leave it for the purpose of this function.